### PR TITLE
TypeScript: Include fragment metadata in selections

### PR DESCRIPTION
**Pull request by gantonious**
# Summary

When generating TypeScript query selections Syrup in the past automatically merged selections from both fragment spreads and inline fragments as part of the current object's selections. It would specify a `typeCondition` on the merged selections to ensure they are only considered on types the fragment applies to. Unfortunately this selection merging leads to a loss in data when used with our iOS Relay client. Where it only explicitly takes into account type condition for fragment and inline fragment selections. This loss in data can lead to attempting to select values that need not be selected from the cache, leading to unexpected cache misses. 

This PR updates our TypeScript selection structure to explicitly include fragment spread and inline fragments as their own type of selection alongside associated metadata such as their type condition. This will allow us to properly select fragment fields when integrating with our Relay client.

Shoutout to @samrmur for the context on this bug.

# Tophat Instructions

Verify CI is ✅ and the updated query structure makes sense.

### CODE REVIEW

File: Sources/SyrupCore/Stencil Extensions/SelectionSetExtension.swift [hunk 1]
```diff
@@ -386,31 +417,12 @@ class SelectionSetExtension: Extension {
...
-			if let fragmentSpread = selection.fragmentSpread {
-				fragmentSpreadRenders.append(fragmentSpread.name)
-			}
-		}
-		
-		var addFragmentSpreadRenders = ""
-		var renderPassedTypeCondition = ""
-		
-		if let typeCondition = typeCondition {
-			renderPassedTypeCondition = ".map(x => copyWithTypeCondition(x, \(typeCondition)))"
```
Comment (gantonious) [inline Sources/SyrupCore/Stencil Extensions/SelectionSetExtension.swift:399]: All this is no longer needed as this info is now encoded as part of the fragment selection.

File: Templates/TypeScript/FragmentBarrel.stencil [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {% include "Helpers/Header.stencil" %}
 {% if fragmentNames.count > 0 %}{% for name in fragmentNames %}export type { {{ name }}FragmentData } from "./{{ name }}"
 export { {{ name|lowercasedFirstLetter }}Selections } from "./{{ name }}"
+export { {{ name|lowercasedFirstLetter }}TypeCondition } from "./{{ name }}"
```
Comment (gantonious) [inline Templates/TypeScript/FragmentBarrel.stencil:4]: Exposed a fragment's `TypeCondition` so we can reference it when constructing a fragment selection for queries referencing this fragment.

File: Sources/SyrupCore/Stencil Extensions/SelectionSetExtension.swift [hunk 3]
```diff
@@ -327,28 +336,50 @@ class SelectionSetExtension: Extension {
...
+				}
+				
+				render.append("\n\(spacing)    selections: \(fragmentSpread.name.lowercasedFirstLetter)Selections")
+				
+				render.append("\n\(spacing)  }")
+				
+				fieldRenders.append(render)
+			} else if let field = selection.field {
+				var render = "\n\(spacing)  {"
+				render.append("\n\(spacing)    selectionType: \"field\",")
```
Comment (gantonious) [inline Sources/SyrupCore/Stencil Extensions/SelectionSetExtension.swift:382]: Thoughts on including the `selectionType` on the standard field selection? We could leave this out and detect fields by checking if `selectionType` is `undefined`. The benefit of doing so is reducing the amount of data we send over the bridge when triggering queries. It feels kind of hacky thought which is why I opted to be explicit here.

**Review by dingua — APPROVED** [review]

File: Sources/SyrupCore/Stencil Extensions/SelectionSetExtension.swift (hunk 3 above)
Comment (frankfka) [reply Sources/SyrupCore/Stencil Extensions/SelectionSetExtension.swift:382]: I think it'd be more straightforward to be explicit 👍🏻

**Review by Caiopia — APPROVED** [review]

**Review by samrmur — APPROVED** [review]

**Pull request merged**